### PR TITLE
TYP: type excel util module

### PR DESCRIPTION
**Pull request by phofl**
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them

### CODE REVIEW

File: pandas/io/excel/_util.py [hunk 1]
```diff
@@ -12,10 +19,15 @@
     is_list_like,
 )
 
-_writers: MutableMapping[str, str] = {}
+if TYPE_CHECKING:
+    from pandas.io.excel._base import ExcelWriter
 
+    ExcelWriter_t = Type[ExcelWriter]
```
Comment (twoertwein) [inline pandas/io/excel/_util.py:25]: Can probably use `type[ExcelWriter]`

File: pandas/io/excel/_util.py [hunk 2]
```diff
@@ -28,10 +40,12 @@ def register_writer(klass):
     if not callable(klass):
         raise ValueError("Can only register callables as engines")
     engine_name = klass.engine
+    # for mypy
+    assert isinstance(engine_name, str)
     _writers[engine_name] = klass
 
 
-def get_default_engine(ext, mode="reader"):
+def get_default_engine(ext: str, mode: str = "reader") -> str:
```
Comment (twoertwein) [inline pandas/io/excel/_util.py:48]: `mode: Literal["reader", "writer"]`

File: pandas/io/excel/_util.py [hunk 3]
```diff
@@ -145,7 +159,31 @@ def _range2cols(areas: str) -> list[int]:
...
+
+
+@overload
+def maybe_convert_usecols(usecols: list[str]) -> list[str]:
+    ...
+
+
+@overload
+def maybe_convert_usecols(
+    usecols: Callable[[Hashable], object]
```
Comment (twoertwein) [inline pandas/io/excel/_util.py:174]: I assume it returns the same callable? Could probably use `TypeVar("something", Callable[[Hashable], object])` to return the same function interface.

File: pandas/io/excel/_util.py [hunk 4]
```diff
@@ -224,7 +274,9 @@ def fill_mi_header(row, control_row):
     return row, control_row
 
 
-def pop_header_name(row, index_col):
+def pop_header_name(
+    row: list[Hashable], index_col: int | list[int]
```
Comment (twoertwein) [inline pandas/io/excel/_util.py:278]: Could also try to use TypeVar here to link the `list[Hashable]`. Probably only useful if the code later requires a more precise type.

**Review by twoertwein — CHANGES_REQUESTED** [review]
Feel free to ignore the TypeVar comments, I'm not sure whether they will be useful

File: pandas/io/excel/_util.py (hunk 1 above)
Comment (phofl) [reply pandas/io/excel/_util.py:25]: Thx, did not know that

File: pandas/io/excel/_util.py (hunk 2 above)
Comment (phofl) [reply pandas/io/excel/_util.py:48]: adjusted

File: pandas/io/excel/_util.py (hunk 3 above)
Comment (phofl) [reply pandas/io/excel/_util.py:174]: Yep makes sense

File: pandas/io/excel/_util.py (hunk 4 above)
Comment (phofl) [reply pandas/io/excel/_util.py:278]: Not sure if this is useful here

**Review by twoertwein — APPROVED** [review]

File: pandas/io/excel/_util.py (hunk 1 above)
Comment (twoertwein) [reply pandas/io/excel/_util.py]: PEP 585 has a nice overview list: https://www.python.org/dev/peps/pep-0585/#implementation

Comment (phofl) [reply pandas/io/excel/_util.py]: Thx, this is handy

**Comment by jreback** [comment]
one more rebase pls

**Comment by phofl** [comment]
Done

File: pandas/io/excel/_util.py [hunk 5]
```diff
@@ -145,7 +161,29 @@ def _range2cols(areas: str) -> list[int]:
...
+    ...
+
+
+@overload
+def maybe_convert_usecols(usecols: None) -> None:
+    ...
+
+
+def maybe_convert_usecols(
+    usecols: str | list[int] | list[str] | usecols_func | None,
```
Comment (jreback) [inline pandas/io/excel/_util.py:185]: prob can share this with the parsers (followon, maybe put in _typing.py)?

**Review by jreback — APPROVED** [review]
lgtm small comment

File: pandas/io/excel/_util.py (hunk 5 above)
Comment (phofl) [reply pandas/io/excel/_util.py:185]: The configs there are a bit different unfortunately. Don't think we can share efficiently

**Pull request merged**